homeassistant/components/matter/media_player.py

```diff
@@ -0,0 +1,137 @@
+        max_level = self.get_matter_attribute_value(
+            clusters.LevelControl.Attributes.MaxLevel
+        )
+        if max_level is None or max_level <= 0:
+            max_level = 254
+
+        if volume == 0:
+            # If volume level is 0, turn off the volume
+            await self.async_mute_volume(True)
+            return
+        # Unmute the volume if it was muted
+        await self.async_mute_volume(False)
+
+        # Convert HA volume level (0-1) to Matter level (0-254)
+        # Matter uses 0-254 for volume levels, where 0 is off
+        matter_volume_level = int(volume * max_level)
+        await self.send_device_command(
+            clusters.LevelControl.Commands.MoveToLevel(level=matter_volume_level)
+        )
+
+    async def async_mute_volume(self, mute: bool) -> None:
+        """Mute or unmute the media player."""
+        # OnOff attribute == True state means volume is on, so HA should show mute switch as off
+        if mute:
+            # Send Matter Off command
+            await self.send_device_command(
+                clusters.OnOff.Commands.Off(),
+            )
+        # OnOff attribute == False means volume is off (muted), so HA should show mute switch as on
+        else:
+            await self.send_device_command(
+                clusters.OnOff.Commands.On(),
+            )
+
+    @callback
+    def _update_from_device(self) -> None:
+        """Update from device."""
+        matter_volume = self.get_matter_attribute_value(
+            clusters.LevelControl.Attributes.CurrentLevel
+        )
+        LOGGER.debug(
+            "matter_volume: %s for %s",
+            matter_volume,
+            self.entity_id,
+        )
+        if matter_volume is None:
+            # Attribute unavailable, cannot determine volume state
+            return
+        if matter_volume == 0:
+            self._attr_is_volume_muted = True
+        # Convert Matter CurrentLevel (0-254) to HA volume level (0-1)
+        else:
+            self._attr_is_volume_muted = False
+            max_level = self.get_matter_attribute_value(
+                clusters.LevelControl.Attributes.MaxLevel
+            )
+            if max_level is None or max_level <= 0:
+                max_level = 254
+            self._attr_volume_level = matter_volume / float(max_level)
+
+
+# Discovery schema(s) to map Matter Attributes to HA entities
+DISCOVERY_SCHEMAS = [
+    MatterDiscoverySchema(
+        platform=Platform.MEDIA_PLAYER,
+        entity_description=MatterMediaPlayerEntityDescription(
+            key="MatterMediaPlayer", name=None
+        ),
+        entity_class=MatterMediaPlayer,
+        required_attributes=(
+            clusters.OnOff.Attributes.OnOff,
+            clusters.LevelControl.Attributes.CurrentLevel,
+        ),
+        optional_attributes=(clusters.LevelControl.Attributes.MaxLevel,),
+        device_type=(device_types.Speaker,),
+    ),
```

Comment:
As we don't `allow_multi=True`, this will remove the Matter [mute-switch entity](https://github.com/home-assistant/core/blob/67aeafa7971cacd21d196513bae9425c8d4ab263/homeassistant/components/matter/switch.py#L249-L266) if a speaker device also supports `CurrentLevel`. This would be considered a breaking change, but I guess there's still no real devices that actually use this?

Reply:
Ah, I see there was already some discussion on this above: https://github.com/home-assistant/core/pull/150470#discussion_r2551155123